xaitk_saliency/impls/perturb_image/rise.py

```diff
@@ -0,0 +1,90 @@
+from xaitk_saliency.interfaces.perturb_image import PerturbImage
+import PIL.Image
+from typing import Optional, Tuple, List, Dict, Any
+import numpy as np
+from skimage.transform import resize
+
+
+class RISEPertubation (PerturbImage):
+    def _generate_masks(self, input_size: Tuple[int, int]) -> np.ndarray:
+        """
+        Randomly crop, upscale, and interpolate binary masks to generate a set
+        of random masks to apply to the image.
+        :param (int, int) input_size:
+            Size of the model's input as (rows, cols). Binary masks are upsampled to
+            this resolution to be applied to (multiplied with) the input later.
+            E.g. (224, 224).
+        """
+        masks = np.empty((self.N, *input_size))
+        cell_size = np.ceil(np.array(input_size) / self.s)
+        # Upscale factor
+        up_size = (self.s + 1) * cell_size
+        for i in range(self.N):
+            # Random shifts
+            x = self.rng.integers(0, cell_size[0])
```

Comment:
>     2. due to random cropping in the upsampling operation, the results of `perturb()` are no longer idempotent

You could initialize a random generator every time within this function when there is a provided seed (or even a separate seed from that used in grid generation) to provide optionally deterministic behavior.

Reply:
This way you also don't have to retain the `self.rng` instance.

Reply:
This will be changed with the streaming update, not going to require any changes here regarding this.